Asynchronous scene viewer

johnhaddon opened the pull request:
This follows on from the work on the [asynchronous image viewer](#2578) by also making the scene viewer asynchronous, so that it updates progressively as computations at a per-location level complete on the background threads.

The approach I've taken is to unify the viewer code with the code used to drive interactive renders. This approach has a few side benefits :

- There's less code to maintain - the two systems had a huge amount of similarity but with subtly different features. Now we just have the RenderController to maintain.
- We could now expand/collapse hierarchies in interactive renders if we wanted. This represents good progress towards using interactive renders as a raytraced viewport.
- The scene graph update has been separated from the GL rendering code. This simplifies future optimisations to the GL rendering.
- The GL renderer has become more capable, making it more suitable as a replacement for the old one in IECoreGL, should we want to use it in the Maya or Houdini viewports for instance.

As well as reviewing, it would be great to get as much interactive testing of this as possible, since it does represent quite a shakeup.

andrewkaufman commented:
I played with this a little bit today while doing some other work. I haven't had any crashes or other inexplicable issues, but the behaviour for a medium-heavy animated caches is quite odd I'd say. I have a private scene I can send if that'd help.

andrewkaufman commented:
I'm also getting this occasionally: `WARNING : IECoreGL::Renderer::option : Unknown option "sampleMotion".`

johnhaddon commented:
> Hey John, I saw you mentioned "This represents good progress towards using interactive renders as a raytraced viewport." - is this opening a way to implement alternative viewports in general, like Hydra or something else?

Potentially. It decouples the "querying the gaffer scene for changes" part of the viewer from the "draw the scene" part of the viewer. So it should make an alternative viewer easier to implement...

johnhaddon commented:
Thanks for the testing and feedback - responses below...

> I'm also getting this occasionally: WARNING : IECoreGL::Renderer::option : Unknown option "sampleMotion".

Fixed in f804b0e.

> I think cameras are sticking around longer than they should be.

Fixed in d9bc30d.

> I've also got a mysterious ortho camera, which the lookThrough tells me is /gaffer:defaultCamera that seems to exist no matter which node I view from.

Fixed in 654f26b.

> I haven't had any crashes or other inexplicable issues, but the behaviour for a medium-heavy animated caches is quite odd I'd say. I have a private scene I can send if that'd help.

Yes please - I'm keen to get this over the line.

johnhaddon commented:
Note to self : we need to update GafferHQ/dependencies to use Cortex 10.0.0-a27 before we can release this.

johnhaddon commented:
I've rebased and pushed some further improvements. The new commits start at 4e8a135, and improve the interaction between progressive updates and animation scrubbing and manipulation. I've also included a few miscellaneous bug fixes for things I encountered along the way, and added bindings for RenderController so I could add a few more unit tests.

I do feel like 2506eca and e5cf73d might be made cleaner with some changes to the Renderer API, but I want to defer that decision till later. I'd lke to consider how it relates to Arnold 5's new IPR API, and the API for a certain other renderer I've been playing with, and it's not yet clear to me what that means...

I think this covers all the comments so far...

johnhaddon commented:
I've got to the bottom of the test failures and will have a fix tomorrow...

johnhaddon commented:
Squashed the fix into the final commit - all other commits are unchanged.

johnhaddon commented:
I've pushed a series of commits (91ee149 - 58e20b2) that fix the look-through sync issue. We now force an immediate update of the look-through camera/light and all its children before drawing, leaving just the rest of the scene to update asynchronously. Would be good to get some feedback on this.

And, of course, just as I thought this was done, I found another issue. This one is a crash caused by the TransformTool's use of Monitors, the registration/deregistration of which is not thread-safe with respect to background processes. I have a hacky fix, but will sleep on it to see if I can come up with something better. Either way, the fix won't impact on the code changes here, so don't let it hold you back from review/testing.

@donboie, @mattigruener, @danieldresser-ie, would be great to get your eyes on this in addition to @andrewkaufman, even if just for testing rather than review.

andrewkaufman commented:
I tried it out on my animated production scene. Jumping to a bespoke frame or even scrubbing is much nicer now. Thanks for that!

However, actually pushing play on the timeline results in essentially a static viewer until I press stop. I presume this is because the playback tries to increment the frame faster than the scene can compute, so every frame is cancelled before it displays? Unless you have an idea how to fix it, I'm inclined to leave it this way at first, since I don't expect many people actually push play on the timeline... I could be wrong though, in which case those users would definitely think that's a bug.

Also, I tried the camera look-through and it seemed fine, but when I looked through an IE light, I got a crash. Not sure if its the same one you mentioned above, I did have the translate manipulator active when I started. I'll send you the stacktrace.

johnhaddon commented:
> Jumping to a bespoke frame or even scrubbing is much nicer now. Thanks for that!

Great!

> I presume this is because the playback tries to increment the frame faster than the scene can compute, so every frame is cancelled before it displays? 

Exactly.

> Unless you have an idea how to fix it, I'm inclined to leave it this way at first

I do have a few ideas, but I'm also inclined to leave it for now. I think it's more important to get the async stuff out for testing.

> When I looked through an IE light, I got a crash.

The stack you sent indicates that this is the same crash I mentioned. I've pushed the fix in 0edd003.

andrewkaufman commented:
Actually that last commit seems to have improved the playback a bit as well.

One more bug for you. Open the same scene I sent you last time. In the HierarchyView expand to `/IE_sunnyBounce/lights/simple_nodecay/filllight_Dir`, then press F in the viewer. The default cam focuses to nothing, and it doesn't seem recoverable. No errors in my terminal. It works as expected in Gaffer 0.47

johnhaddon commented:
> One more bug for you. Open the same scene I sent you last time. In the HierarchyView expand to /IE_sunnyBounce/lights/simple_nodecay/filllight_Dir, then press F in the viewer.

Thanks for catching that - fixed in e0042ab. I've also rebased the earlier commits to fix a conflict with master.
